cli/pkg/local/app.go

```diff
@@ -578,6 +588,79 @@ func (a *App) emitStartEvent(ctx context.Context) error {
 	return nil
 }
 
+// initiateAuthFlow starts the OAuth2 PKCE flow to authenticate the user and get a rill access token.
+func (a *App) initiateAuthFlow(httpPort int, secure bool) http.Handler {
+	scheme := "http"
+	if secure {
+		scheme = "https"
+	}
+	redirectURL := fmt.Sprintf("%s://localhost:%d/auth/callback", scheme, httpPort)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// generate random state
+		b := make([]byte, 32)
+		_, err := rand.Read(b)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to generate state: %s", err), http.StatusInternalServerError)
+			return
+		}
+		state := base64.StdEncoding.EncodeToString(b)
+
+		authenticator, err := pkce.NewAuthenticator(a.adminURL, redirectURL, database.AuthClientIDRillLocal, state)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to generate pkce authenticator: %s", err), http.StatusInternalServerError)
+			return
+		}
+		a.pkceAuthenticators[state] = authenticator
+		authURL := authenticator.GetAuthURL(state)
+		http.Redirect(w, r, authURL, http.StatusFound)
+	})
+}
+
+// handleAuthCallback handles the OAuth2 PKCE callback to exchange the authorization code for a rill access token.
+func (a *App) handleAuthCallback() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		code := r.URL.Query().Get("code")
+		if code == "" {
+			http.Error(w, "missing code", http.StatusBadRequest)
+			return
+		}
+		state := r.URL.Query().Get("state")
+		if code == "" {
+			http.Error(w, "missing state", http.StatusBadRequest)
+			return
+		}
+
+		authenticator, ok := a.pkceAuthenticators[state]
+		if !ok {
+			http.Error(w, "invalid state", http.StatusBadRequest)
+			return
+		}
+
+		// remove authenticator from map
+		delete(a.pkceAuthenticators, state)
+
+		if authenticator == nil {
+			http.Error(w, "failed to get authenticator", http.StatusInternalServerError)
+			return
+		}
+
+		// Exchange the code for an access token
+		token, err := authenticator.ExchangeCodeForToken(code)
+		if err != nil {
+			http.Error(w, "failed to exchange code for token", http.StatusInternalServerError)
+			return
+		}
+		// save token and redirect to home screen or TODO url provided in initial request i.e. initiateAuthFlow
+		err = dotrill.SetAccessToken(token)
+		if err != nil {
+			http.Error(w, "failed to save access token", http.StatusInternalServerError)
+			return
+		}
```

Comment:
In addition to updating `dotrill`, we may also need to set `ch.AdminTokenDefault = token` (this is used in the current process for subsequent requests – `.rill` is only read at start time). But not sure if `ch` is in scope here (from the CLI), we may need to pass it in.